jni/lib/src/util/jlist.dart

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: overridden_fields
+
+import 'dart:collection';
+import 'dart:ffi';
+
+import 'package:jni/src/util/jset.dart';
+
+import '../accessors.dart';
+import '../jni.dart';
+import '../jobject.dart';
+import '../jvalues.dart';
+import '../third_party/generated_bindings.dart';
+import '../types.dart';
+import 'jiterator.dart';
+
+class JListType<$E extends JObject> extends JObjType<JList<$E>> {
+  final JObjType<$E> E;
+
+  const JListType(
+    this.E,
+  );
+
+  @override
+  String get signature => r"Ljava/util/List;";
+
+  @override
+  JList<$E> fromRef(JObjectPtr ref) => JList.fromRef(E, ref);
+
+  @override
+  JObjType get superType => const JObjectType();
+
+  @override
+  final superCount = 1;
+
+  @override
+  int get hashCode => Object.hash(JListType, E);
+
+  @override
+  bool operator ==(Object other) {
+    return other.runtimeType == JListType && other is JListType && E == other.E;
+  }
+}
+
+class JList<$E extends JObject> extends JObject with ListMixin<$E> {
+  @override
+  late final JObjType $type = type(E);
+
+  final JObjType<$E> E;
+
+  JList.fromRef(
+    this.E,
+    JObjectPtr ref,
+  ) : super.fromRef(ref);
+
+  static final _class = Jni.findJClass(r"java/util/List");
+
+  /// The type which includes information such as the signature of this class.
+  static JListType<$E> type<$E extends JObject>(
+    JObjType<$E> E,
+  ) {
+    return JListType(
+      E,
+    );
+  }
+
+  static final _arrayListClassRef = Jni.findJClass(r"java/util/ArrayList");
+  static final _ctorId = Jni.accessors
+      .getMethodIDOf(_arrayListClassRef.reference, r"<init>", r"()V");
+  JList.array(this.E)
+      : super.fromRef(Jni.accessors.newObjectWithArgs(
+            _arrayListClassRef.reference, _ctorId, []).object);
+
+  static final _sizeId =
+      Jni.accessors.getMethodIDOf(_class.reference, r"size", r"()I");
+  @override
+  int get length => Jni.accessors
+      .callMethodWithArgs(reference, _sizeId, JniCallType.intType, []).integer;
+
+  @override
+  set length(int newLength) {
+    RangeError.checkNotNegative(newLength);
+    while (length < newLength) {
+      add(E.fromRef(nullptr));
+    }
+    while (newLength < length) {
+      removeAt(length - 1);
```

Comment:
A high level question is whether you want Java list to behave like dart list here, or as java `List`.

Here we are not hiding implementation detail of JNI but interface of java itself.

@dcharkes wdyt?

Reply:
We had a prior discussion about this and we decided to make them behave more like Dart lists. One of the reasonings being that you can always generate your own `java.lang.List` which is more Java like using `jnigen`.